### PR TITLE
Rework term indexing and stuck rewrites

### PR DESCRIPTION
**Pull request by geo2a**
Closes https://github.com/runtimeverification/hs-backend-booster/issues/457
* do not assign term index `None` to unrecognized terms, use `Anything` instead
* do not abort when there are no rules, get stuck instead

TODO:
- [x] add unit test that assigns `Anything` to `dotk()` and gets stuck
- [x] ~Investigate why several KEVM tests fail (see list in the comment below)~ -- resolved itself with merging main.
- [x] run Kontrol performance tests
- [x] run KEVM performance tests

### CODE REVIEW

**Comment by geo2a** [comment]
A bunch of KEVM tests fail:
```
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/dsvalue-read-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/dsvalue-read-pass-summarize-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/end-subuu-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/flipper-addu48u48-fail-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/flipper-bids-pass-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/flipper-tau-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/flipper-ttl-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/flopper-cage-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/flopper-file-pass-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/functional-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-addui-fail-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-addui-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-deny-diff-fail-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-move-diff-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-mului-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[functional/evm-int-simplifications-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-muluu-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-slip-pass-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-subui-fail-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-subui-pass-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/vat-subui-pass-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[examples/erc20-spec.md]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/cat-exhaustiveness-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[examples/solidity-code-spec.md]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/cat-file-addr-pass-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[mcd/dai-adduu-fail-rough-spec.k]
FAILED src/tests/integration/test_prove.py::test_pyk_prove[examples/storage-spec.md]
```

**Comment by geo2a** [comment]
| Test                                     | georgy-drop-rule-aborts time | master-5976645 time | (georgy-drop-rule-aborts/master-5976645) time
|------------------------------------------|------------------------------|---------------------|-----------------------------------------------
| mcd/flipper-ttl-pass-spec.k              | 85.24                        | 89.7                | 0.9502787068004458
| mcd/vat-subui-pass-spec.k                | 138.67                       | 143.59              | 0.9657357754718294
| bihu/forwardToHotWallet-success-1-spec.k | 29.37                        | 28.28               | 1.0385431400282885
| mcd/flipper-tau-pass-spec.k              | 95.37                        | 91.19               | 1.0458383594692402
| mcd/vat-subui-pass-rough-spec.k          | 140.54                       | 133.66              | 1.0514738889720185
| TOTAL                                    | 489.18999999999994           | 486.4200000000001   | 1.005694667160067

File: library/Booster/Pattern/Rewrite.hs [hunk 1]
```diff
@@ -124,8 +124,6 @@ rewriteStep cutLabels terminalLabels pat = do
                     then idxRules
                     else Map.unionWith (<>) idxRules anyRules
 
-    when (null rules) $ throw (NoRulesForTerm pat.term)
```
Comment (jberthold) [inline library/Booster/Pattern/Rewrite.hs:127]: We should remove the `NoRulesForTerm` constructor (further below in this file).

**Review by jberthold — APPROVED** [review]
LGTM, except mopping up the `NoRulesForTerm` constructor

**Pull request merged**